WIP: Use a n-dimensional Bresenham algorithm in draw_nd

marcotama opened the pull request:
## Description

This is a possible enhancement of the `draw_nd` function that uses the Bresenham algorithm (extended to n dimensions).
I suppose its strongest advantage is that it is unaffected by floating point error.

The current proposed version is vectorized in the number of dimensions, while a Python loop iterates over the various points. I suppose the original implementation is faster in most use cases although I have not run any tests.

The current algorithm is a WIP because it works exclusively on integers (as did the original Bresenham. Ideas for how to handle floats are welcome (I am currently truncating them).

My idea I have is to use the decimal value to initialize the errors, but to do that I think I need to find the least common denominator of said decimals after converting them to fractions, which has two problems:
    a) it only works with fractional values (ok, on a computer no really real numbers exist, but then the LCD could have a huge denominator resulting in huge initial errors (think plotting from `(0,0)` to `(cos(30°), sin(30°))`, and
    b) computing the LCD sounds like extra work one might not want to do (but I guess we have the `integer` flag for that?).


## Checklist

<!-- It's fine to submit PRs which are a work in progress! -->
<!-- But before they are merged, all PRs should provide: -->
- [ ] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [ ] Benchmark in `./benchmarks`, if your changes aren't covered by an
  existing benchmark
- [ ] Unit tests
- [ ] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)

<!-- For detailed information on these and other aspects see -->
<!-- the scikit-image contribution guidelines. -->
<!-- https://scikit-image.org/docs/dev/contribute.html -->

## For reviewers

<!-- Don't remove the checklist below. -->
- [ ] Check that the PR title is short, concise, and will make sense 1 year
  later.
- [ ] Check that new functions are imported in corresponding `__init__.py`.
- [ ] Check that new features, API changes, and deprecations are mentioned in
      `doc/release/release_dev.rst`.
- [ ] Consider backporting the PR with `@meeseeksdev backport to v0.14.x`

marcotama committed:
Add draw_nd almost-equivalent which uses Bresenham algorithm

pep8speaks commented:
Hello @marcotama! Thanks for opening this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:

* In the file [`skimage/draw/draw_nd.py`](https://github.com/scikit-image/scikit-image/blob/96c83ad49d4479476416449cbbd3ae59caf860be/skimage/draw/draw_nd.py):

> [Line 110:1](https://github.com/scikit-image/scikit-image/blob/96c83ad49d4479476416449cbbd3ae59caf860be/skimage/draw/draw_nd.py#L110): [E302](https://duckduckgo.com/?q=pep8%20E302) expected 2 blank lines, found 1
> [Line 173:17](https://github.com/scikit-image/scikit-image/blob/96c83ad49d4479476416449cbbd3ae59caf860be/skimage/draw/draw_nd.py#L173): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 182:1](https://github.com/scikit-image/scikit-image/blob/96c83ad49d4479476416449cbbd3ae59caf860be/skimage/draw/draw_nd.py#L182): [E305](https://duckduckgo.com/?q=pep8%20E305) expected 2 blank lines after class or function definition, found 1
> [Line 191:12](https://github.com/scikit-image/scikit-image/blob/96c83ad49d4479476416449cbbd3ae59caf860be/skimage/draw/draw_nd.py#L191): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 197:80](https://github.com/scikit-image/scikit-image/blob/96c83ad49d4479476416449cbbd3ae59caf860be/skimage/draw/draw_nd.py#L197): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (92 > 79 characters)

jni commented:
Hey @marcotama! This is great! @stefanv was not super happy that we didn't get this right in #2043. There's a bit of cleanup to be done before merging, but in principle we totally want this in!

I'm sorry that I don't have time to do a full review right now (I'm not actually familiar with the Bresenham algorithm), but the way forward would be to replace the original `draw_line_nd` function with this one, *once* the inner loop is sped up by either vectorising or switching to Cython.

In other news, I notice you're in Melbourne??? We should meet up and finish this up in person! =D You can email me at juan.nunez-iglesias@monash.edu

marcotama committed:
Implement Cython version of the N-dimensional Bresenham algorithm

marcotama commented:
I implemented a Cython version of the same algorithm, but I am not an expert, and I am pretty sure it can be optimized further. It's obviously faster than the original one, but a fair comparison would have Cython version of both.

Here are the results of a simple test; as expected, the `linspace` based algorithm outperforms  Bresenham on few dimensions, in the Python form.
![comparison](https://user-images.githubusercontent.com/1052500/68537955-f4917900-03c0-11ea-96d7-7d16f83c2ac5.png)

@jni I sent you an email :)

jni commented:
@marcotama awesome! Can you produce the same plots for just 3D/4D and line length of 16, 32, ..., 16384? I think that is a much more common use case than 100-D data. :joy:

marcotama commented:
Obviously I wasn't thinking fourth-dimensionally! Silly me!

Here you go!

![2dim](https://user-images.githubusercontent.com/1052500/68538485-3d4d3000-03c9-11ea-89ac-18f6002034d2.png)
![3dim](https://user-images.githubusercontent.com/1052500/68538484-3d4d3000-03c9-11ea-9de3-ee083e6bfe2f.png)
![4dim](https://user-images.githubusercontent.com/1052500/68538483-3d4d3000-03c9-11ea-8963-177df4cf51d9.png)

marcotama commented:
(ignore the lines, they are linear interpolations)

jni committed:
Initial working version of nD Bresenham Cython

jni committed:
Slightly more accurate...

jni commented:
@marcotama I pushed our changes from today to your branch.

@scikit-image/core this is still WIP and not ready for review... Some of our adaptations of the algorithm to floating point values are not working correctly.

marcotama committed:
Almost there

marcotama committed:
Numbers seem correct now

marcotama committed:
Add more tests and add explanations as comments

marcotama committed:
Remove bankers rounding - it's reason d'eitre is to maintain the energy in a system by sometimes rounding up and sometimes down, but in the case of graphics it results in jitter if a line shifts by 1px

marcotama commented:
@jni Can you have a look at the current status? I believed I fixed the math, and the results are now almost identical to the existing implementation. (Cleanup remains a todo)

I say "almost" because there are differences due to a couple of reasons:
- I removed the banker's rounding (i.e. rounding to the nearest even number) for the following reason: banker's rounding was first introduced (afaik) to maintain (on average) the energy in "the system". By sometimes rounding `~.5` up and sometimes rounding it down you get that effect. However, in a graphic library I put forward the case that that should not be the policy. If that were the case, a line that passes at some `~.5` and is shifted up 1px at a time in an animation will show jitter if banker's rounding is used. This should not be the case.
- Different algorithms have different biases for cases when the line passes exactly between two pixels (regardless whether the bias is explicit, like when using rounding, or implicit, due to e.g. >/>= in the code on negative/positive numbers).

Obviously this needs to be double and triple checked ;)

marcotama commented:
I have written a little notebook to generate some plots to compare the two algorithms. Here are the plots of the test cases in the code (I made the 3D to 2D).

![(0, 0) to (2, 2)](https://user-images.githubusercontent.com/1052500/69908937-d3afc700-1446-11ea-82f7-c84fd36344a8.png)
![(0, 0) to (3, -5)](https://user-images.githubusercontent.com/1052500/69908938-d3afc700-1446-11ea-930a-c82426520887.png)
![(0, 0) to (4, 2)](https://user-images.githubusercontent.com/1052500/69908939-d3afc700-1446-11ea-93cb-d2d4da665add.png)
![(0, 0) to (5, -3)](https://user-images.githubusercontent.com/1052500/69908940-d4485d80-1446-11ea-984f-8b565f9822ed.png)
![(1, 0) to (5, -1 5)](https://user-images.githubusercontent.com/1052500/69908941-d4485d80-1446-11ea-943b-75f2874a341b.png)
![(1, 1) to (3, 3)](https://user-images.githubusercontent.com/1052500/69908942-d4e0f400-1446-11ea-96d5-687e9a7020f0.png)
![(1, 1) to (3, 5)](https://user-images.githubusercontent.com/1052500/69908943-d4e0f400-1446-11ea-986d-c7fba019f595.png)
![(1, 1) to (5, -2 4)](https://user-images.githubusercontent.com/1052500/69908944-d4e0f400-1446-11ea-9a07-f12c55287baa.png)
![(1, 1) to (5, -2 5)](https://user-images.githubusercontent.com/1052500/69908945-d5798a80-1446-11ea-8e03-899d66574809.png)
![(1, 1) to (5, -2 6)](https://user-images.githubusercontent.com/1052500/69908946-d5798a80-1446-11ea-847b-758b2cbc6c9a.png)
![(1, 1) to (5, 2 5)](https://user-images.githubusercontent.com/1052500/69908947-d5798a80-1446-11ea-8476-84dfa8c0a445.png)
![(1, 1) to (5, 3)](https://user-images.githubusercontent.com/1052500/69908948-d6122100-1446-11ea-8c7e-12bc457cf2fc.png)
![(1 1, 1 4) to (5 3, 2 2)](https://user-images.githubusercontent.com/1052500/69908949-d6122100-1446-11ea-923c-85f0e86f7a01.png)

marcotama commented:
The code I used is:
```python
from skimage.draw.draw_nd import line_nd, my_line_nd
import matplotlib.pyplot as plt
from matplotlib.collections import PatchCollection
from matplotlib.patches import Rectangle

tests = [
    ((0, 0), (2, 2)), # m=1 (x0 is even)
    ((1, 1), (3, 3)), # m=1 (x0 is odd)
    ((0, 0), (4, 2)), # m=2 (x0 is even)
    ((1, 1), (5, 3)), # m=2 (x0 is odd)
    ((1, 1), (3, 5)), # x is the second coordinate
    ((1, 1), (5, 2.5)), # y1 is decimal
    ((1.1, 1.4), (5.3, 2.2)), # all decimals (this produces a bug - a repeated point in the original implementation, due to the call to np.ceil (maybe a -0.5 is due in there))
    ((0, 0), (5, -3)), # m<0
    ((0, 0), (3, -5)), # m<0, x is the second coordinate
    ((1, 1), (5, 2.5)), # 3 dimensions
    ((1, 1), (5, -2.5)), # 3 dimensions, m>0 and m<0
    ((1, 0), (5, -1.5)), # 3 dimensions, m>0 and m<0, shifting causes different shape in the old implementation (compared to previous test)
    ((1, 1), (5, -2.4)), # just need to tease out why the results are what they are in this and the next example
    ((1, 1), (5, -2.6))
]

def plot_pixels(ax, pixels, line):
    facecolor='k'
    edgecolor='k'
    alpha=0.5
    x0, y0, x1, y1 = line
        
    ax.grid()
    ax.axes.set_aspect('equal', 'datalim')
    squares = []
    xs, ys = pixels
    for x, y in zip(xs, ys):
        square = Rectangle((x-.5, y-.5), 1, 1)
        squares.append(square)
    pc = PatchCollection(squares, facecolor=facecolor, alpha=alpha, edgecolor=edgecolor)
    ax.add_collection(pc)
    if abs(x1-x0) > abs(y1-y0):
        ax.set_xlim(min(xs) - 1, max(xs) + 1)
    else:
        ax.set_ylim(min(ys) - 1, max(ys) + 1)
    
    ax.plot([x0, x1], [y0, y1], color='k')
    
    

for start, stop in tests:
    fig, (ax1, ax2) = plt.subplots(nrows=1, ncols=2, figsize=(10,5))
    fig.suptitle(f"{start} to {stop}")
    line = start[0], start[1], stop[0], stop[1]
    
    ax1.set_title("Original")
    pixels1 = line_nd(start, stop, endpoint=True)
    plot_pixels(ax1, pixels1, line)
    
    ax2.set_title("New")
    pixels2 = my_line_nd(start, stop, endpoint=True)
    plot_pixels(ax2, pixels2, line)
    
    fig.savefig(f"{start} to {stop}.png")
```

jni commented:
@marcotama can you explain

```
    ((1, 1), (5, 2.5)), # 3 dimensions
```
? That is clearly two dimensions. =D

Thanks for the code and images, though, this is super thorough and easy to introspect! I love it! Gonna have a proper look in the next few days.

jni commented:
@marcotama looking at the examples again, this kinda hits the nail on the head for me:

> I aesthetically prefer the new algorithm, but the old algorithm seems to have less error

Is this caused by the choice of rounding or something else? It makes me tic a little. =P

>  These two show that the two algorithms have a different bias for rounding. There is not such thing as "correct".

See above.

I'd like to understand why the error is higher in the new version, as I don't really see why that should be the case. But, other than that, I love these diagnostics and am convinced that this is a better algorithm for this purpose. Thanks for the work so far @marcotama! It's great! I don't have time to review the code thoroughly right now but I hope to be able to do so later this week. 

I will say that eventually we want your code to completely replace my existing code, but happy to keep it there while we keep diagnosing this.

stefanv commented:
It would be fun to do the animation that you hinted at above; I wonder if it will show interesting artifacts.

It seems like the old algorithm has a tendency to be more "symmetric", which intuitively feels the way it should be; any intuition why?  I'm also surprised that the error is different between the two—what could cause that?

matthew-brett commented:
We have changed the base branch for this PR from the `main` branch to the `pre-megamove-integration` branch.

## Summary - what this means for your PR

You can continue to work on your PR, now based on `pre-megamove-integration`.  You should update your branch to the current state of `pre-megamove-integration` (this is what the Github "Update branch" button does).   Please **do not merge the scikit-image `main` branch into your PR branch**.

Once we've merged your PR into the `pre-megamove-integration` branch, we'll update your changes for Megamove, and merge into the `main` branch.

In addition - our automated script detected that your branch had merge conflicts with the pre-Megamove branch. Please make sure to merge in the `pre-megamove-integration` branch, and resolve the merge conflicts in the way that you want.

Read on for the Megamove rationale, and more detailed instructions.

<details>

## Rationale

We've changed the base branch to make it easier and smoother to continue to work on your PR, while we work through a rather large change in the directory structure of the repository - a change that we have called *Megamove*.

See [the Megamove explanation issue](https://github.com/scikit-image/scikit-image/issues/8232) for more detail.

Changing the base branch means you do not see merge conflicts related to the *Megamove* change, and can continue to work with the repository in its original state, without worrying about porting those changes across to the new post-Megamove structure.

As you'll see in the linked issue, you can continue to work on your PR as before, and you can continue to use the Github GUI to "Update branch" - this updates from the new `pre-megamove-integration` branch.   However, there is one caveat: please, do not try and merge the `scikit-image` `main` branch into this one, but instead, keep track of the `pre-megamove-integration` branch.  The issue explains more.

To emphasize again, **do not try and merge the scikit-image `main` branch**.  You will get Megamove related merge conflicts that are rather tricky to resolve correctly.

When we've approved your PR, we'll merge into the `pre-megamove-integration` branch, and then process it so we can merge it into the `main` branch.   We'll let you know when we've merged your PR changes into `main`.

</details>

Any questions, please reply here.
